modules/currency-exchange/src/lib.rs

```diff
@@ -0,0 +1,403 @@
+
+use codec::Encode;
+use frame_support::{decl_error, decl_module, decl_storage, ensure, fail, Parameter};
+use primitives::exchange::{
+	Airdrop, CurrencyConverter, Error as ExchangeError, MaybeLockFundsTransaction, RecipientsMap,
+};
+use sp_runtime::DispatchResult;
+use sp_std::vec::Vec;
+
+/// Peer blockhain interface.
+pub trait Blockchain {
+	/// Block hash type.
+	type BlockHash: Parameter;
+	/// Transaction type.
+	type Transaction: Parameter;
+	/// Transaction inclusion proof type.
+	type TransactionInclusionProof: Parameter;
+
+	/// Verify that transaction is a part of given block.
+	fn verify_transaction_inclusion_proof(
+		transaction: &Self::Transaction,
+		block: Self::BlockHash,
+		proof: &Self::TransactionInclusionProof,
+	) -> bool;
+}
+
+/// The module configuration trait
+pub trait Trait: frame_system::Trait {
+	/// Peer blockchain type.
+	type PeerBlockchain: Blockchain;
+	/// Peer blockchain transaction parser.
+	type PeerMaybeLockFundsTransaction: MaybeLockFundsTransaction<
+		Transaction = <Self::PeerBlockchain as Blockchain>::Transaction,
+	>;
+	/// Map between blockchains recipients.
+	type RecipientsMap: RecipientsMap<
+		PeerRecipient = <Self::PeerMaybeLockFundsTransaction as MaybeLockFundsTransaction>::Recipient,
+		Recipient = Self::AccountId,
+	>;
+	/// This blockchain currency amount type.
+	type Amount;
+	/// Converter from peer blockchain currency type into current blockchain currency type.
+	type CurrencyConverter: CurrencyConverter<
+		SourceAmount = <Self::PeerMaybeLockFundsTransaction as MaybeLockFundsTransaction>::Amount,
+		TargetAmount = Self::Amount,
+	>;
+	/// Something that could grant money.
+	type Airdrop: Airdrop<Recipient = Self::AccountId, Amount = Self::Amount>;
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		/// Invalid peer blockchain transaction provided.
+		InvalidTransaction,
+		/// Peer transaction has invalid amount.
+		InvalidAmount,
+		/// Peer transaction has invalid recipient.
+		InvalidRecipient,
+		/// Cannot map from peer recipient to this blockchain recipient.
+		FailedToMapRecipients,
+		/// Failed to convert from peer blockchain currency to this blockhain currency.
+		FailedToCovertCurrency,
+		/// Airdrop has failed.
+		AirdropFailed,
+		/// Transaction is not finalized.
+		UnfinalizedTransaction,
+		/// Transaction funds are already claimed.
+		AlreadyClaimed,
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		/// Imports lock fund transaction of the peer blockchain.
+		#[weight = 0] // TODO: update me (https://github.com/paritytech/parity-bridges-common/issues/78)
+		pub fn import_peer_transaction(
+			origin,
+			transaction: <<T as Trait>::PeerBlockchain as Blockchain>::Transaction,
+			block: <<T as Trait>::PeerBlockchain as Blockchain>::BlockHash,
+			proof: <<T as Trait>::PeerBlockchain as Blockchain>::TransactionInclusionProof,
+		) -> DispatchResult {
+			frame_system::ensure_signed(origin)?;
```

Comment:
Long-term we should reward the ones how submit such transactions. Can we have something super simplistic for now (like a `OnTransactionImport(who)` hook in the trait definition) so that it can be plugged in when needed without changing the code?

Reply:
`trait OnTransactionSubmitted`